### DISCUSSION TITLE
Amber API process

### DISCUSSION THREAD

**Discussion by petebitt99** [amberelectric/public-api#196] [Q&A] [answered]
I am just starting trying to use the Amber API and have a few questions.
I have generated the API token - will I be able to continue to use this key or will it change over time?
I have used the key to authorise API calls and run some tests. I got my site ID doing this.  Is the site ID static? Or can it change?
I intend to call the API's using Homey (homey.app) which is already integrated to my Tesla Gateway, so with the info from Amber I will be able to build some logic into a Dashboard which will either notify me of certain conditions, or allow me to control my load.
I'm hoping I can get this going and then contribute / test the Amber API so that I can develop further automations.

**Comment by rob0101** [comment]
Welcome.
The key and the site ID are static.

**Reply by petebitt99** [reply]
Thank you.  I'm working my way through this so I appreciate your support.
My next question is what do I need to do to run the API outside of the Amber website.  If I try and call the API (https://api.amber.com.au/v1/sites/01H92N0GC70BAZHM48P2C52J3M/prices/current?next=5&previous=1&resolution=30) from my browser i get {"message":"Unauthorized"}

**Reply by madpilot** [reply]
You can't really run it in a browser, because there is no way to add authentication headers - you can, however use cURL or use a programming language to access it.

**Reply by petebitt99** [reply]
Ok, I will have to knock up some test code in Homey.
So do I need to do this
curl -X 'GET' \
  'https://api.amber.com.au/v1/sites' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer psk_f31a4b0b3ee67b8db63067a7d0385bd0'
 before I make the call for current prices?
If the answer is yes advice on where I can go to work out how to use cURL would be appreciated.

**Reply by madpilot** [reply]
That call will return your site id, so you don't really need to do it before every call - you can do it once and then save it. The only time you need to do it again is if you leave Amber and sign back up (a new site is added to the results).

You should just be able to cut and paste that in to your terminal. What operating system are you on?

**Reply by petebitt99** [reply]
It is a home automation hub (https://homey.app/en-au/blog/a-technical-introduction-homey/) which has an ARM processor running Linux. It is Java script based and allows me to write what they call Homey script from which I would call the API.
Example Homey script:
// Function to shuffle an array
function shuffleArray(arr) {
  let newArr = arr.slice();
  for (let i = newArr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));

**Reply by madpilot** [reply, marked as answer]
Looks like it's just a straight up Node-based JavaScript script.

Something like this should get you started:

```javascript
const apiKey = 'psk_f31a4b0b3ee67b8db63067a7d0385bd0'
const siteId = '01H92N0GC70BAZHM48P2C52J3M'

const response = await fetch(`https://api.amber.com.au/v1/sites/${siteId}/prices/current`, { 
  headers: { 
    'Authorization': `Bearer ${apiKey}` 
  } 
}
const prices = await response.json()
const generalPrice = prices.find((price) => price.channelType === 'general')?.perKwh
const feedInPrice = prices.find((price) => price.channelType === 'feedIn')?.perKwh
```

**Reply by petebitt99** [reply]
Thank you.  There was a missing ); just before const prices. Other than that it worked perfectly. Thanks again for your help.

**Reply by petebitt99** [reply]
Great, that's all working now and I can get notifications when the FIT is negative and my Battery SoC is >95% etc.

Looking at the API descriptions it looks like there is no way to find out programmatically when Smartshift canges state from Consume Battery.  Can you confirm please.

Smartshift regulary puts my system into the Preserve state, meaning I draw from the grid, even though there is no need for me to do so. Unless I'm missing something.  I would like to get notifications of these shifts in state.

**Reply by madpilot** [reply]
Coming soon

**Reply by petebitt99** [reply]
I don't suppose the "coming soon" might include the days sun outlook?

**Reply by madpilot** [reply]
Do you mean the solar forecast? Yeah, most likely.

**Reply by petebitt99** [reply]
That would be handy.

**Reply by madpilot** [reply]
Have a look at Solcast. They have a hobby account which is free and their forecasts are much better than ours.

**Reply by petebitt99** [reply]
Looks like the current prices API has changed. Now next number of intervals instead of dates as input.

**Reply by madpilot** [reply]
The current price API has never used dates as an input - it's always been numbers?

**Reply by petebitt99** [reply]
Cool, I have enrolled in SolCast.  My Homey Pro had a SolCast App which exposes the SolCast data. It works well.

**Reply by petebitt99** [reply]
That's weird. Only a few days ago it asked for dates and I was using this:
`https://api.amber.com.au/v1/sites/${siteId}/prices?startDate=${startDate}&endDate=${endDate}&resolution=30`
and it was working.

**Reply by madpilot** [reply]
There are two different end points - the /prices endpoint you are describing there.

The _current_ prices endpoint is /prices/current which takes next and previous attributes.

Are you sure you aren't getting the two confused?

**Reply by petebitt99** [reply]
Aah, yes I see now. Sorry Myles

**Reply by madpilot** [reply]
As an aside, you _probably_ want the `/prices/current` endpoint for what you want to do

**Comment by petebitt99** [comment]
Yes I agree. It is the better option. Although with the javascript I can make both work.  I am slowly putting together the components I need for a solar management system. I can now get how much solar is expected to be produced over the next 3 hours, the expected charge level of my battery in that time, and for how many minutes into the future the negative Fit is expected to last.  Now I need to create the logic for when and at what level I throttle the inverter.
Although I know nothing about HA I suspect that if I was using it this has already been worked out.

**Reply by madpilot** [reply]
The main reason you want `/prices/current` not `/prices` is with the latter at 5 minutes to midnight, you only have one forecasted price to work with, where as `/prices/current` will always return _n_ forecasts in to the future.
